File: 198-house-robber/198-house-robber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
		int n = nums.size();
		if (n == 1) return nums[0];
		int i2 =  nums[0];
		int i1 = max(nums[0], nums[1]);

		for (int i = 2 ; i < n; i++ ) {
            int temp = i1;
			i1 = max(i2 + nums[i], i1);
            i2 = temp;
		}
		return max(i1,i2);
	}
};

File: 198-house-robber/198-house-robber.py
class Solution:
    
    def recur(self,nums,i,n,arr):
        if(i>(n-1)):
            return 0
        
        if(arr[i]!=-1):
            return arr[i]
        
        # 1 2 3 1
          
        op1 = nums[i] + self.recur(nums,i+2,n,arr)
        op2 = self.recur(nums,i+1,n,arr)
        arr[i] = max(op1,op2)
        return arr[i]
        
    def rob(self, nums: List[int]) -> int:
        arr = [-1]*101
        n = len(nums)
        ans = self.recur(nums,0,n,arr)
        return ans
        